File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.cpp
class Solution {
  public:
    int longestSubarray(vector<int>& arr, int k) {
        // code here
        map<int,int> mp; // one is to store the suma nd other is tor= store the index
        
        int sum =0;
        int ans =0;
        for(int i=0;i<arr.size();i++){
            sum += arr[i];
            if(sum == k) ans = max(ans,i+1);
            if(mp.find(sum-k)!=mp.end()){
                ans = max(ans,(i-mp[sum-k]));
            }
            
            if(mp.find(sum)==mp.end()){
                mp[sum] = i;
            }
            
        }
        return ans;
    }
};

File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.py
class Solution:
    def longestSubarray(self, arr, k):  
        # code here
        mp = {}
        sum =0
        ans=0;
        for i in range(0,len(arr)):
            sum+=arr[i]
            if sum == k:
                ans = max(ans,i+1)
            if sum-k in mp:
                ans = max(ans,i-mp[sum-k])
            if sum not in mp:
                mp[sum]=i;
        return ans
        
    
